File: src/strategies/narrative-momentum.ts
/**
 * Narrative Momentum Strategy
 *
 * Thesis: Crypto markets move in narratives (AI tokens, DeFi, RWA, L2s, etc.).
 * When a narrative catches fire — measured by correlated social velocity across
 * multiple tokens in the same sector — the whole sector tends to run.
 *
 * This strategy:
 *  1. Groups tokens into narratives/sectors
 *  2. Tracks social mention velocity for each sector
 *  3. When a sector's velocity spikes >narrativeVelocityThreshold×, find the
 *     sector laggard (lowest price appreciation vs sector average) and enter long
 *
 * Why laggard: sector leaders have already pumped; the laggard hasn't "caught up" yet.
 * This mimics how retail rotates within a hot sector.
 */

import type { TradeSignal, ScannerConfig } from '../types.js';
import { randomUUID } from 'crypto';

export type NarrativeId =
  | 'ai_tokens'
  | 'defi_bluechip'
  | 'layer2'
  | 'rwa'
  | 'gaming_metaverse'
  | 'meme'
  | 'depin'
  | 'liquid_staking'
  | 'btc_ecosystem'
  | 'privacy';

export const NARRATIVE_MEMBERS: Record<NarrativeId, string[]> = {
  ai_tokens:        ['FET', 'AGIX', 'OCEAN', 'RENDER', 'TAO', 'WLD'],
  defi_bluechip:    ['UNI', 'AAVE', 'CRV', 'MKR', 'SNX', 'COMP'],
  layer2:           ['ARB', 'OP', 'MATIC', 'IMX', 'STRK', 'MANTA'],
  rwa:              ['ONDO', 'POLYX', 'CFG', 'TRU', 'MPL'],
  gaming_metaverse: ['AXS', 'SAND', 'MANA', 'GALA', 'ILV', 'YGG'],
  meme:             ['DOGE', 'SHIB', 'PEPE', 'WIF', 'BONK', 'FLOKI'],
  depin:            ['HNT', 'MOBILE', 'IOT', 'RNDR', 'FIL', 'AR'],
  liquid_staking:   ['LDO', 'RPL', 'SFRXETH', 'ANKR', 'PENDLE'],
  btc_ecosystem:    ['STX', 'ORDI', 'SATS', 'RATS', 'MUBI'],
  privacy:          ['XMR', 'ZEC', 'SCRT', 'DUSK', 'CTXC'],
};

interface NarrativeState {
  id: NarrativeId;
  socialVelocity: number;       // current velocity (relative to 30d baseline)
  baselineVelocity: number;
  memberPriceChanges: Map<string, number>; // symbol → 24h price change pct
  lastUpdated: number;
}

const narrativeStates = new Map<NarrativeId, NarrativeState>();

export function updateNarrativeSocialData(
  narrativeId: NarrativeId,
  currentMentions: number,
  baselineMentions: number,
): void {
  const state = narrativeStates.get(narrativeId) ?? {
    id: narrativeId,
    socialVelocity: 1,
    baselineVelocity: baselineMentions,
    memberPriceChanges: new Map(),
    lastUpdated: Date.now(),
  };
  state.socialVelocity = baselineMentions > 0 ? currentMentions / baselineMentions : 1;
  state.baselineVelocity = baselineMentions;
  state.lastUpdated = Date.now();
  narrativeStates.set(narrativeId, state);
}

export function updateNarrativeMemberPrice(narrativeId: NarrativeId, symbol: string, priceChangePct: number): void {
  const state = narrativeStates.get(narrativeId);
  if (state) {
    state.memberPriceChanges.set(symbol, priceChangePct);
  }
}

function findLaggard(state: NarrativeState, productIdMap: Map<string, string>): {
  symbol: string; productId: string; priceChangePct: number;
} | null {
  let laggard: { symbol: string; productId: string; priceChangePct: number } | null = null;

  for (const [sym, pct] of state.memberPriceChanges.entries()) {
    const pid = productIdMap.get(sym);
    if (!pid) continue;
    if (!laggard || pct < laggard.priceChangePct) {
      laggard = { symbol: sym, productId: pid, priceChangePct: pct };
    }
  }

  return laggard;
}

export function scanNarrativeMomentum(
  productIdMap: Map<string, string>, // symbol → productId (only available tokens)
  config: ScannerConfig,
  currentPrices: Map<string, number>,
): TradeSignal | null {
  let bestSignal: TradeSignal | null = null;
  let bestScore = 0;

  for (const [narrativeId, state] of narrativeStates.entries()) {
    if (state.socialVelocity < config.narrativeVelocityThreshold) continue;
    if (Date.now() - state.lastUpdated > 1_800_000) continue; // stale data

    const laggard = findLaggard(state, productIdMap);
    if (!laggard) continue;

    const currentPrice = currentPrices.get(laggard.symbol);
    if (!currentPrice) continue;

    // Score: higher velocity = higher score; laggard being more negative = more upside
    const velocityScore = Math.min(35, (state.socialVelocity - config.narrativeVelocityThreshold) * 10);
    const laggardScore = Math.min(20, Math.max(0, -laggard.priceChangePct * 100));
    const score = Math.min(88, 48 + velocityScore + laggardScore);

    if (score > bestScore) {
      bestScore = score;
      bestSignal = {
        id: randomUUID(),
        symbol: laggard.symbol,
        productId: laggard.productId,
        strategy: 'narrative_momentum',
        side: 'long',
        tier: 'swing',
        score,
        confidence: score > 72 ? 'medium' : 'low',
        sources: ['social'],
        reasoning: `${narrativeId.replace(/_/g, ' ')} narrative velocity ${state.socialVelocity.toFixed(1)}× baseline; ${laggard.symbol} lagging sector by ${(laggard.priceChangePct * 100).toFixed(1)}% — rotation play`,
        entryPrice: currentPrice,
        stopPrice: currentPrice * 0.94,
        suggestedSizeUsd: 80,
        expiresAt: Date.now() + 7_200_000, // 2h
        createdAt: Date.now(),
      };
    }
  }

  return bestSignal;
}

File: src/strategies/narrative_momentum.py
"""Narrative Momentum Strategy — sector rotation play."""

import threading
import time
import uuid
from dataclasses import dataclass, field
from typing import Optional

from src.types import TradeSignal, ScannerConfig, MarketContext

NarrativeId = str

NARRATIVE_MEMBERS: dict[str, list[str]] = {
    "ai_tokens":        ["FET", "AGIX", "OCEAN", "RENDER", "TAO", "WLD"],
    "defi_bluechip":    ["UNI", "AAVE", "CRV", "MKR", "SNX", "COMP"],
    "layer2":           ["ARB", "OP", "MATIC", "IMX", "STRK", "MANTA"],
    "rwa":              ["ONDO", "POLYX", "CFG", "TRU", "MPL"],
    "gaming_metaverse": ["AXS", "SAND", "MANA", "GALA", "ILV", "YGG"],
    "meme":             ["DOGE", "SHIB", "PEPE", "WIF", "BONK", "FLOKI"],
    "depin":            ["HNT", "MOBILE", "IOT", "RNDR", "FIL", "AR"],
    "liquid_staking":   ["LDO", "RPL", "SFRXETH", "ANKR", "PENDLE"],
    "btc_ecosystem":    ["STX", "ORDI", "SATS", "RATS", "MUBI"],
    "privacy":          ["XMR", "ZEC", "SCRT", "DUSK", "CTXC"],
}


@dataclass
class NarrativeState:
    id: str
    social_velocity: float = 1.0
    baseline_velocity: float = 0
    member_price_changes: dict[str, float] = field(default_factory=dict)
    last_updated: float = 0


_narrative_states: dict[str, NarrativeState] = {}
_lock = threading.Lock()
_MAX_NARRATIVE_STATES = 500

STRATEGY_META = {
    "strategies": [
        {"id": "narrative_momentum", "function": "scan_narrative_momentum",
         "description": "Detects narratives gaining momentum and trades laggard catch-up",
         "tier": "swing"},
    ],
    "signal_sources": ["social", "news"],
}


def _evict_lru_if_needed() -> None:
    """Evict least-recently-updated entries when _narrative_states exceeds max size."""
    if len(_narrative_states) <= _MAX_NARRATIVE_STATES:
        return
    sorted_keys = sorted(_narrative_states, key=lambda k: _narrative_states[k].last_updated)
    to_remove = len(_narrative_states) - _MAX_NARRATIVE_STATES
    for k in sorted_keys[:to_remove]:
        del _narrative_states[k]


def update_narrative_social_data(narrative_id: str, current_mentions: int, baseline_mentions: int) -> None:
    with _lock:
        if narrative_id not in _narrative_states:
            _narrative_states[narrative_id] = NarrativeState(id=narrative_id)
        _evict_lru_if_needed()
        state = _narrative_states[narrative_id]
    state.social_velocity = current_mentions / baseline_mentions if baseline_mentions > 0 else 1
    state.baseline_velocity = baseline_mentions
    state.last_updated = time.time() * 1000


def update_narrative_member_price(narrative_id: str, symbol: str, price_change_pct: float) -> None:
    with _lock:
        state = _narrative_states.get(narrative_id)
    if state:
        state.member_price_changes[symbol] = price_change_pct


def _find_laggard(state: NarrativeState, product_id_map: dict[str, str]) -> Optional[dict]:
    laggard = None
    for sym, pct in state.member_price_changes.items():
        pid = product_id_map.get(sym)
        if not pid:
            continue
        if laggard is None or pct < laggard["price_change_pct"]:
            laggard = {"symbol": sym, "product_id": pid, "price_change_pct": pct}
    return laggard


def scan_narrative_momentum(
    product_id_map: dict[str, str],
    config: ScannerConfig,
    current_prices: dict[str, float],
    ctx: Optional[MarketContext] = None,
) -> Optional[TradeSignal]:
    now = time.time() * 1000
    best_signal = None
    best_score = 0

    # Regime-adjust velocity threshold: require stronger signal in bear markets
    velocity_threshold = config.narrative_velocity_threshold
    if ctx and ctx.phase == "bear":
        velocity_threshold *= 1.5

    with _lock:
        snapshot = list(_narrative_states.items())

    for narrative_id, state in snapshot:
        if state.social_velocity < velocity_threshold:
            continue
        if now - state.last_updated > 1_800_000:
            continue

        laggard = _find_laggard(state, product_id_map)
        if not laggard:
            continue
        current_price = current_prices.get(laggard["symbol"])
        if not current_price:
            continue

        velocity_score = min(35, (state.social_velocity - velocity_threshold) * 10)
        laggard_score = min(20, max(0, -laggard["price_change_pct"] * 100))
        score = min(88, 48 + velocity_score + laggard_score)

        if score > best_score:
            best_score = score
            best_signal = TradeSignal(
                id=str(uuid.uuid4()), symbol=laggard["symbol"],
                product_id=laggard["product_id"],
                strategy="narrative_momentum", side="long", tier="swing", score=score,
                confidence="medium" if score > 72 else "low",
                sources=["social"],
                reasoning=f"{narrative_id.replace('_', ' ')} narrative velocity {state.social_velocity:.1f}x; {laggard['symbol']} lagging by {laggard['price_change_pct']*100:.1f}%",
                entry_price=current_price, stop_price=current_price * 0.94,
                suggested_size_usd=80,
                expires_at=now + 7_200_000, created_at=now,
            )

    return best_signal
